contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+The controller will continue to attempt the update to the spec while there is a difference between the current size and size in updated spec.
+If an update is partially successful, user can know this from the logs and attempt to rectify the situation by submitting new updates (that can restore original size(s) or go for a size feasible on all the nodes). 
+
+The policy for vertical scaling can itself be mutable.
+It is desirable to submit a change in policy, confirm its acceptance (by re-reading the changed pod spec) and then submit a change to a resource size impacted by that policy. 
+
+# Design and Implementation
+
+This section highlights some of implementation details by describing API changes and the workflow of vertical scaling on a pod.
+
+## Changes on API and key components
+
+This section describes briefly API changes for pod-level vertical scaling and the related changes on the 'API server', the 'Scheduler', and the 'Kubelet'.
+
+* **ResizeRequest**
+
+A new data structure, ResizeRequest, is added to v1.PodSpec:
+
+```go
+const (
+    ResizeRequested ResizeStatus = "Requested"
+    ResizeAccepted ResizeStatus = “Accepted”
+    ResizeRejected ResizeStatus = “Rejected”
+    ResizeNone      ResizeStatus = "None"
+)
+
+type ResizeRequest struct {
+    RequestStatus ResizeStatus
+    NewResources  []ResourceRequirements // indexed by containers’ index
+}
+
+Type PodSpec {
+    …
+    ResizeRequest ResizeRequest
+    ...
+}
+```
+
+ResizeRequest has two variables, RequestStatus and NewResources.
+RequestStatus represents the status of a resource resizing request.
+ResizeRequested indicates resource resizing for a pod is requested.
+ResizedAccepted and ResizedRejected means that the requested resource resizing is accepted and rejected, respectively, by the Scheduler.
+The NewResources is an array indexed by a container’s index and its each entry holds new resource requirements of a container that needs to resize.
+
+Given a new PodSpec with new resource requirements from a client, first the 'API server' validates it.
+If it is valid, the 'API server' sets the RequestStatus to ResizedRequested and copies the new resource requirements of each container into the NewResources.
+Also, the 'API server' restores the resource requirements of each container of the PodSpec to the original and writes the revised PodSpec to ETCD to communicate with the Scheduler.
+This is because at this moment the PodSpec on ETCD shouldn’t be updated with new resource requirements.
+
+For a pod with ResizeRequested, the 'Scheduler' checks if the node on which the pod currently runs has enough resources to resize the pod.
+The Scheduler notify the 'API server' of the result via 'Resizing' API operation, which will be describe below.
+
+* **Resizing**
+
+A new API, Resizing, for the 'scheduler' is introduced:
+
+```go
+// Resizing resizes the resources allocated to a pod
+type Resizing struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+    Request ResizeRequest
```

Comment:
I guess we should follow spec+status approach in all api objects.

Reply:
for me, I would regard `Resizing` as kind of action not resource, a little bit tricky to define this struct.

Reply:
and we may build command like `kubectl resize`?

Reply:
As @Adohe mentioned, "Resizing" is a kind of action(?) like Binding, which has the same fields.

Reply:
We think that resizing is just another pod spec update, so prefer to utilize the current podspec update methods like kubectl apply, patch, etc. instead of introducing a new kubectl command.

Reply:
> We think that resizing is just another pod spec update, so prefer to utilize the current podspec update methods like kubectl apply, patch, etc. instead of introducing a new kubectl command.

I agree with this

currently, the fact that a node is running a pod implies it has accepted the resource requests the pod made.

this is proposing introducing a state in which a node is running a pod with one set of resources, but may or may not have accepted the pod's currently requested resources. that seems to imply:
* resources should become part of the pod/container status
* when spec....resources do not match status...resources, then the node has not seen/accepted/adjusted to the newly requested resources
* the node should have some way in status to indicate it has seen, but has refused to accept, the newly requested resources, and why

Reply:
@liggitt I would think that container status should include current resources, still don't understand why CRI `ContainerStatus` definition not include container resources.